Allow integers in scientific format in `INSERT JSON`

avelanarius opened the pull request:
Add support for specifing integers in scientific format (for example 1.234e8) in INSERT JSON statement:

```
INSERT INTO table JSON '{"int_column": 1e7}';
```

Before the JSON parsing library was switched to RapidJSON from JsonCpp, this statement used to work correctly, because JsonCpp transparently casts double to integer value.

Inserting a floating-point number ending with .0 is allowed, as the fractional part is zero. Non-zero fractional part (for example 12.34) is disallowed. A new test is added to test all those behaviors.

This behavior differs from Cassandra, which disallows those types of numbers (1e7, 123.0 and 12.34), however some users rely on that behavior and JSON specification itself does not distinct between floating-point numbers and integer numbers (only a single "number" type is defined).

This PR also fixes two minor issues I noticed while looking at the code: wrong blob validation and missing `IsString()` checks that could result in assertion error.

Fixes #10100
Fixes #10114
Fixes #10115

psarna commented on cql3/type_json.cc:
```diff
@@ -187,7 +187,63 @@ struct from_json_object_visitor {
+                out = fmt::format("{}", value.GetUint64());
+            } else if (value.IsDouble()) {
+                // Formatting style of JsonCpp, which
+                // cannot be easily replicated without printf.
```
Did any particular user also rely on that? I'm asking because now, on the other hand, people might already have started to rely on the rjson output, and we'll never stop this back-and-forth. And a huge advantage of the rjson way is that it doesn't take over 20 lines of careful char processing and pointer arithmetic to get the desired result. Unless there's a good reason to restore the previous format, I'd prefer to avoid open-coding anything.

avelanarius replied:
This code handles such a case:
```
INSERT INTO t JSON '{"string_column": 12.34}';
```

and this doesn't work currently (with rjson), so no one is relying on rjson string formatting of floating-point numbers.

avelanarius replied:
However, I'm not aware of anyone relying on old formatting of floating point numbers. Therefore `t.from_string(rjson::print(value))` would be OK to use here instead of manually serializing to string here, but there's some problem with worse precision when using `rjson::print`.

avelanarius replied:
Fixed in amended commit by just using `rjson::print`

avelanarius commented on cql3/type_json.cc:
```diff
@@ -216,10 +216,30 @@ struct from_json_object_visitor {
+    bytes operator()(const simple_date_type_impl& t) { 
+        if (!value.IsString()) {
+            throw marshal_exception(format("{} must be represented as string in JSON", value));
+        }
```
Fixed in the amended commit by introducing `validated_to_string_view` helper method.

nyh commented:
> Please note that I agree that the current behavior (not accepting floating-point values to integer column) is correct. It even turns out that Cassandra 4.0 has the same behavior (not allowing it), however some users ([scylladb/scylla-enterprise#2119](https://github.com/scylladb/scylla-enterprise/issues/2119)) rely on that old casting behavior (and it may be possible that other JSON producers/JavaScript applications also exhibit the same behavior).

Before reviewing this code, I'm worried about this paragraph.

You are describing a case where both Scylla and Cassandra had some old behavior that was changed in some version to a new behavior, and now a user wants to revert it to the old behavior. While reverting to the old behavior will indeed improve compatibility with old versions, it will break compatibility with more recent versions of both Cassandra and Scylla. In 5 years, nobody will remember the old behavior that Scylla had 10 years prior and why we reverted to it, but everyone can complain that Scylla's behavior in this case is non-nonsensical and not compatible with Cassandra, so I think this might be a step in the wrong direction. But you can convince me otherwise - I don't have strong feelings here (yet).

nyh commented on cql3/type_json.cc:
```diff
@@ -194,7 +194,16 @@ struct from_json_object_visitor {
+        if (value.IsString()) {
+            return t.from_string(rjson::to_string_view(value)); 
+        } else if (value.IsBool() || value.IsNumber()) {
+            return t.from_string(rjson::print(value));
```
How does rjson::print() print floating point numbers? Does it have the full accuracy?
If we accept this patch, we should have a test for this.

nyh replied:
I saw later that you do have a test for this (with some strange results showing slightly different last digit in the two versions, but I guess it's meaningless).

avelanarius replied:
RapidJSON uses its own floating-point printing code (https://github.com/Tencent/rapidjson/blob/e4bde977440d4a00f820b6586899e48a972d2493/include/rapidjson/internal/dtoa.h). By default it seems to have quite a long maximum precision configured: `static const int kDefaultMaxDecimalPlaces = 324;`, however it does use scientific notation for large numbers.

In the testing code, I compared the output of it to the previous Scylla 4.0.0 code with JsonCpp. It looks very similar, but for some cases, it printed numbers less precise compared to JsonCpp (despite the same number of digits).

Here is discussion with @psarna about using `rjson::print` instead of manually printing floats: https://github.com/scylladb/scylla/pull/10101#discussion_r810143380

avelanarius replied:
@nyh "Does it have the full accuracy? If we accept this patch, we should have a test for this." What type of test would you add to the existing tests? Some large number? (i.e. `1e200`?)

nyh replied:
No, accuracy is not about the range of exponent (like 200 in your example) but the digits of precision.
For floating-point the accuracy is less important (because I don't think we make any guarantees about it...) but for integers, we can't loose any digit of precision. I gave in a comment below the case of the highest bigint - 9223372036854775808, which has (one) too many digits of accuracy for a double, so if we write 9223372036854775808E0 and set it into a bigint, we can find ourselves with a slightly different integer, like 9223372036854775807. I suggested below we should have a test for this (even if it will still fail in the current implementation).

nyh commented on test/boost/json_cql_query_test.cc:
```diff
@@ -495,6 +496,83 @@ SEASTAR_TEST_CASE(test_insert_json_implicit_casts) {
+                "pk": 3, "v2": 834658736487.7474
+            }'
+        )");
+        cquery_nofail(e, R"(
```
I think shouldn't be just "nofail", it should check the output for being correct...

avelanarius replied:
I don't understand: `INSERT JSON` does not have any output. The inserted row is checked below in `require_rows`.

nyh replied:
What I really meant to say was that we should probably read after the write, to check what was written... But it's not critical - it will only make the test better. It's already better than nothing.

avelanarius replied:
I still don't understand: "we should probably read after the write, to check what was written" - I do perform that read in `require_rows(e, "SELECT pk, v2 FROM implicit_casts", ...`.

nyh replied:
Ok, I missed the read, sorry.

nyh commented on test/boost/json_cql_query_test.cc:
```diff
@@ -495,6 +496,83 @@ SEASTAR_TEST_CASE(test_insert_json_implicit_casts) {
+            {int32_type->decompose(1), utf8_type->decompose("15")},
+
+            // Those are better than the output of Scylla 4.0.0 with JsonCpp:
+            // matching the value in INSERT JSON exactly.
```
This just goes to show how futile it is to decide our behavior based on some unintentional behavior in Scylla 4.0.0. What does Cassandra do here? What do we think is the **right** thing to do here? :-(

avelanarius replied:
Cassandra fails to cast numbers to string:
```
INSERT INTO t JSON '{"pk": 2, "v": 123.32423}';
InvalidRequest: Error from server: code=2200 [Invalid query] message="Error decoding JSON value for v: Expected a UTF-8 string, but got a Double: 123.32423"
```

nyh commented on test/cql-pytest/test_json.py:
```diff
@@ -58,15 +58,15 @@ def test_fromjson_wrong_type_unprepared(cql, table1):
         cql.execute(f"INSERT INTO {table1} (p, v) VALUES ({p}, fromJson('\"dog\"'))")
     with pytest.raises(FunctionFailure):
-        cql.execute(f"INSERT INTO {table1} (p, a) VALUES ({p}, fromJson('3'))")
+        cql.execute(f"INSERT INTO {table1} (p, a) VALUES ({p}, fromJson('[3]'))")
```
Again, you should be proudly keeping the test which demonstrates a difference between Scylla and Cassandra (after splitting it from the part that is the same) - not replacing it with some other random test that doesn't fail... (in the previous case you changed it to an object, here a list).

nyh reviewed (changes requested):
In general, I'm not happy with the idea of reverting our behavior to some undocumented and perhaps even unintentional behavior we had in older versions, which is not compatible with Cassandra. But I'm not strongly against it either - if the powers that be decide that this is the way to go for some reason, then we can do it.

What I am against, however, is removing the tests which demonstrates this difference between Scylla and Cassandra. I think it's the opposite - we need to keep those tests so as never forget that this difference exists. We also need to document this behavior (and also the difference) in our documentation.

Finally, this series has a lot of different things in it. It has some obvious bug fixes, and it has *two different* types of "casting" changes - one about silently casting floating point to integers (potentially losing accuracy!!), and the other one about casting numbers (but not other types...) to strings.

I think all these changes are separate, need to be separately justified, and should be in separate patches and perhaps not even in one series (we could commit them separately, they aren't really related).

By the way, after reading this whole series with its various different fixes, I find myself wondering - what was the missing casting that the actual user complained about? Was it the number-to-string casting? The floating-point-to-int casting? Both?

avikivity commented on cql3/type_json.cc:
```diff
@@ -189,7 +189,7 @@ struct from_json_object_visitor {
         }
         std::string_view string_v = rjson::to_string_view(value);
-        if (string_v.size() < 2 && string_v[0] != '0' && string_v[1] != 'x') {
+        if (string_v.size() < 2 || string_v[0] != '0' || string_v[1] != 'x') {
```
Please file a separate issue for this, and mark it fixed in this commit.

avikivity commented on cql3/type_json.cc:
```diff
@@ -84,6 +84,13 @@ static int64_t to_int64_t(const rjson::value& value) {
+        throw marshal_exception(format("'{}' must be represented as string in JSON", v));
+    }
+    return rjson::to_string_view(v); 
+}
```
Separate issue for this too.

avikivity commented on cql3/type_json.cc:
```diff
@@ -78,8 +78,12 @@ static int64_t to_int64_t(const rjson::value& value) {
+    } else if (value.IsUint64()) {
         return value.GetUint64(); //NOTICE: large uint64_t values will get overflown
+    } else if (value.IsDouble()) {
+        return value.GetDouble();
```
I agree. What happens with DynamoDB and MongoDB? They are schemaless, but if you declare something as a sort key, it suddenly has a type. Although I see DynamoDB has just a number type, not separate integer and float types.

nyh replied:
Right - DynamoDB indeed, like JSON, has just one "number" type so there is no separate notions of integer and floating point. But Scylla does have types, and the above assignment will assign 2, not 2.5, into the column, which sounds to me like a bad idea, and unlikely to be something that a real user will want to do deliberately - and if it's done by mistake an error will be a better result.

nyh replied:
Well, I'm trying to go for "right", not just bug compatibility with older versions. Especially since only one user complained that this changed - and they only cared about integers in scientific notation - not about floating point like 2.5 working.

Your point about 2.000000001 is a very good point. Not because this example (which is clearly NOT a whole number) and not because of any 32-bit integer - the 32-bit integers can all be accurately represented in a double because double has more digits. However, we would have a bug with very high bigint (64-bit integers): The highest bigint is 2^63 - 1 = 9223372036854775807. If this is ever saved as a double, the last digit of precision will not fit and will not be reproduced correctly  :-( This means that if we write for example 9223372036854775.807E3 and RapidJson parses it as a double, when we save it into a bigint we can loose one digit of precision. This is a bug. We'll need to remember (e.g., in the form of a failing test) that we have this bug. Maybe we even need to recognize this case and error-out  (double-to-integer conversion when the double is larger than some threshold) so the user doesn't loose accuracy.

But in more "reasonable" cases, like the one asked for the original user (with numbers around 1E9), this will be 100% accurate.

To be 100% accurate in every case we'll need to fix RapidJson to use unlimited precision when parsing numbers. This is what we do in Alternator. But it's not how RapidJson works...

avikivity commented on cql3/type_json.cc:
```diff
@@ -78,8 +78,12 @@ static int64_t to_int64_t(const rjson::value& value) {
-    } else if (value.GetUint64()) {
+    } else if (value.IsUint64()) {
         return value.GetUint64(); //NOTICE: large uint64_t values will get overflown
+    } else if (value.IsDouble()) {
```
Please remove references to the enterprise repository from open-source commits and the cover letter.

avikivity commented:
I'd like to understand what's the accepted practice. Many many tools use json as a serialization format. Let's look at some and see if they accept floats where an integer is expected.

nyh commented:
> I'd like to understand what's the accepted practice. Many many tools use json as a serialization format. Let's look at some and see if they accept floats where an integer is expected.

Note that JSON doesn't have the distinction between integer and floats - both are just numbers. It is only the RapidJSON library which tries to make this distinction - so for example if the number is "1234" it saves it as an integer, not a double, so the application doesn't lose any accuracy. With that said, "1e7" is a valid JSON number, and arguably **it is** an integer, so Scylla **should** allow an integer to be assigned "1e7" which is as much as integer as "100" - even if C (and apparently RapidJSON) conventions suggest that 1e7 is considered "floating point".

This is why I asked what was the actual user complaint. If the user complained about not being able to set an integer to 1e7, we can fix that, without allowing an integer to be set to 2.5 (with accuracy lost) in the process!

I personally think it's a mistake to allow assigning 2.5 into an integer without any sort of warning, especially if Cassandra doesn't allow it. But assigning 1e7 into an integer **should** can be allowed (if that's what the user needs), and if Cassandra doesn't support it we can argue it's a Cassandra bug.

Your series also deals with a second kind of assignment - assigning a number into a string. This is a completely separate issue, is this also something the user complained about? I think the result is not pretty (why would an application want to assign a number into a string and not write the number as a string in the first place? why does this work only on numbers and not other things like lists or objects?), but I'm less opposed to it - but I'd still like to see a bit more motivation on why this sort of converting assignments are useful, desirable, Cassandra-compatible (?), or what.

nyh commented:
I sent a patch to the mailing list, titled "[PATCH v2] test/cql-pytest: reproducer for JSON scientific-notation integer problem", which what I think the user was trying to do, and what should be fixed - and all the rest should be fixed in separate patches with separate justifications.

The point is, the user didn't have "floating point" numbers like 1.2. They had a number like `1.23456789E+9` which is just another way to write the **integer** 1234567890, so should have been accepted as an integer, even if RapidJson (and Cassandra) initially read it into a double after having seen the "." or "E". We don't need to accept **any** floating point number like 1.2 to accept whole numbers written in scientific notation.

avelanarius commented on cql3/type_json.cc:
```diff
@@ -78,8 +78,12 @@ static int64_t to_int64_t(const rjson::value& value) {
     } else if (value.IsUint()) {
         return value.GetUint();
-    } else if (value.GetUint64()) {
+    } else if (value.IsUint64()) {
```
The implication of this bug was that if the user provided a floating-point value where an integer value was required, `GetUint64()` would get executed (in the `if` condition) without checking `IsUint64()` that resulted in RapidJSON assertion firing off:

```
ServerError: An unexpected error occurred server side on /172.30.222.11:9042: JSON error: condition not met: data_.f.flags & kUint64Flag
```

avelanarius commented on cql3/type_json.cc:
```diff
@@ -194,7 +194,16 @@ struct from_json_object_visitor {
+        } else if (value.IsBool() || value.IsNumber()) {
+            return t.from_string(rjson::print(value));
+        } else {
+            // We don't serialize objects, arrays.
```
This is based on behavior of Scylla 4.0.0 with JsonCpp.

avelanarius commented:
> I'd like to understand what's the accepted practice. Many many tools use json as a serialization format. Let's look at some and see if they accept floats where an integer is expected.

Looking at PostgreSQL (I will check others too):

JSONB type - it automatically changes `1e7` to `10000000`:

```
postgres=# CREATE TABLE t2 (p int, v jsonb, CONSTRAINT p_pkey PRIMARY KEY (p));
CREATE TABLE
postgres=# INSERT INTO t2(p, v) VALUES (1, '{"v": 1234}');
INSERT 0 1
postgres=# INSERT INTO t2(p, v) VALUES (2, '{"v": 1e7}');
INSERT 0 1
postgres=# SELECT v->'v' FROM t2;
 ?column? 
----------
 1234
 10000000
(2 rows)

postgres=# SELECT * FROM t2;
 p |        v        
---+-----------------
 1 | {"v": 1234}
 2 | {"v": 10000000}
(2 rows)

postgres=# SELECT (v->'v')::int4 FROM t2;
   int4   
----------
     1234
 10000000
(2 rows)
```

JSONB type with floating-point value - `::int4` truncates the fractional part:
```
postgres=# INSERT INTO t2(p, v) VALUES (3, '{"v": 1.364}');
INSERT 0 1
postgres=# SELECT v->'v' FROM t2;
 ?column? 
----------
 1234
 10000000
 1.364
(3 rows)

postgres=# SELECT (v->'v')::int4 FROM t2;
   int4   
----------
     1234
 10000000
        1
(3 rows)
```

`INSERT JSON` equivalent in Postgres - fails to insert `1e7` to `int` column:
```
CREATE TABLE t (p int, b int, CONSTRAINT p_pkey PRIMARY KEY (p));
INSERT INTO t SELECT p, b FROM json_populate_record(NULL::t, '{"p": 1, "b": 1e7}');
ERROR:  invalid input syntax for type integer: "1e7"
```

avelanarius commented:
When Jackson Java library handles deserialization to an object, if floating-point value is provided for integer field, it transparently casts float to int (1e6 to 1000000, 1.234 to 1):

```java
public static class TestClass {
    public long val;
}

public static void main(String[] args) {
    String json = "{\"val\":1e6}";
    String json2 = "{\"val\":1.234}";

    try {
        TestClass testItem = new ObjectMapper().readValue(json, TestClass.class);
        System.out.println(testItem);

        TestClass testItem2 = new ObjectMapper().readValue(json2, TestClass.class);
        System.out.println(testItem2);

        // Prints:
        // TestClass{val=1000000}
        // TestClass{val=1}
    } catch (JsonProcessingException e) {
        e.printStackTrace();
        // Does NOT throw!
    }
}
```

nyh commented:
> JSONB type - it automatically changes `1e7` to `10000000`:
> 
> ```
> postgres=# CREATE TABLE t2 (p int, v jsonb, CONSTRAINT p_pkey PRIMARY KEY (p));
> CREATE TABLE
> postgres=# INSERT INTO t2(p, v) VALUES (1, '{"v": 1234}');
> INSERT 0 1
> postgres=# INSERT INTO t2(p, v) VALUES (2, '{"v": 1e7}');
> INSERT 0 1
> postgres=# SELECT v->'v' FROM t2;
>  ?column? 
> ----------
>  1234
>  10000000

I think JSONB is different from what we're doing. It's just a way to store the entire JSON you specify, and indeed "1e7" and "10000000" are equivalent JSON (both refer to the same number) so it's ok if it changes one to the other.

> postgres=# SELECT (v->'v')::int4 FROM t2;

I think this is different from what we're doing. The user explicitly asks here to case something into an int, so it's reasonable that floating point is truncated when the user explicitly asked for it.

> `INSERT JSON` equivalent in Postgres - fails to insert `1e7` to `int` column:
> 
> ```
> CREATE TABLE t (p int, b int, CONSTRAINT p_pkey PRIMARY KEY (p));
> INSERT INTO t SELECT p, b FROM json_populate_record(NULL::t, '{"p": 1, "b": 1e7}');
> ERROR:  invalid input syntax for type integer: "1e7"
> ```

Ok, this is similar to our current behavior, and I consider it a bug too. It is perfectly understandable how someone with decades of C programming experience will take a look at "1e7" or "123.4e1" and say that this is not an integer because it looks like a float constant, although it is a whole number.
On the other hand, one might ask what possessed a user who decided to write 123.4e1 instead of 1234. There is no real good reason to do that...

To summarize my view is that we have two good options: One to tell the user "this is not a bug" - the user should output 1234 instead of 123.4e1 and everything will work. The second option is to allow 123.4e1 and other scientific-notation integers, but NOT non-whole numbers like 123.4. I don't think the third option - of truncating any floating-point number - is a good one.

avelanarius commented:
> Your series also deals with a second kind of assignment - assigning a number into a string. This is a completely separate issue, is this also something the user complained about?

Yes, assigning a number into a string is not something that the user encountered (only the "floating-point to integer" case). I fixed it proactively, just in case someone relies on that behavior (in older Scylla with JsonCpp it worked). In general, I only added those types of casts that worked in the previous Scylla with JsonCpp, nothing more.

Whether this kind of assignment should be also fixed is a similar judgment call as the "floating-point to integer" case.

avelanarius commented:
v2:
- Reduced this PR: now only a case of using scientific notation for integers is fixed. Cases of "implicit casting" from boolean to integer types, from non-text types to text types are no longer fixed by this PR, as those problems were not encountered by the user, I personally would not consider them bugs and they was less consensus it should be fixed.
- Before casting double to int64, I checked whether the fractional part is zero and otherwise throw error.
- Fixed error message in `validated_to_string_view` to contain the type name
- Added tests of passing non-text type to simple_date_type_impl, time_type_impl, uuid_type_impl, inet_addr_type_impl (disallowed)
- Added test of passing 2^53-1 as float (maximum integer guaranteed by JSON spec to be "interoperable")
- Removed references to enterprise issue
- Separate "Fixed" issues for two first commits
- cql-pytest test are no longer modified (they work correctly as-is, rebased this PR on top of latest master)

nyh commented on cql3/type_json.cc:
```diff
@@ -78,8 +78,35 @@ static int64_t to_int64_t(const rjson::value& value) {
+        // Check if the value contains disallowed fractional part (.34 from 12.34).
+        // With RapidJSON and an integer value in range [-(2^53)+1, (2^53)-1], 
+        // the fractional part will be zero as the entire value
+        // fits in 53-bit significand. RapidJSON's parsing code does lose accuracy:
```
this "does" is a typo, should be **doesn't**, maybe?

avelanarius replied:
Yes, this is a typo.

nyh reviewed (approved):
Very good. I just think I found one typo in a comment, maybe?

I'll take care of the test/cql-pytest.py (dropping the xfail from now-passing tests) myself later.

slivne commented:
this was merged 24acabc2bdcace36f7682ff7efeac3501974f254

its part of 2021.1.10 and later releases

@tomer-sandler: given the scope of the issue they had - downtime - Nexar should spin a test cluster and validate it actually works for them - and not only reply on us.

tomer-sandler commented:
Roger that. @lucmargui is going to communicate that to them.

avikivity commented:
Do not describe customer issues on this open-source tracker.
